### ISSUE TITLE
Restarting the mom kills the running job(s)

### ISSUE THREAD

**Issue by itkovian** [adaptivecomputing/torque#324]
We have built new RPMs to deal with the XML issue in the job scripts. However when restarting the mom on the nodes, the running jobs are (sometimes) killed.

We have reason to believe the culprit can be found in the comparison between job_start_time and session_start_time in scan_non_child_tasks (in mom_mach.c). At some point this is executed and the following comparison is made:

~~~~
if(job_start_time == session_start_time)
  {
  found = 1;
  }
~~~~

I think that only way to move found tasks to the exited state is by not setting found to 1. On occasions this seems to happen, but not always.

Drilling down, the time used comes from /proc/<pid>/stat, for both job_start_time and sessions_start_time, but in case of clock skew these can be different.

To force reproduction of the issue, you can force the btime values to be different, e.g., 

~~~~
service ptpd2 stop # or ntpd 
hwclock --date 'Thu Sep 10 14:56:56 CEST 2015' --set 
hwclock -s 
~~~~

In that case, the running processes will be terminated when the pbs_mom service is restarted.

Would it be feasible to not have an _exact_ comparison, but to allow for some clock skew to occur?

We have built from the 5.1-dev branch, on commit 71f6ede, see #322.

**Comment by stdweird** [comment]
some more info:
googling around shows that btime was static in 2.4 kernels, but since 2.6 it's just `gettimeofday() - uptime`
uptime (1st filed from /proc/uptime) can be considered correct (it's based on jiffies), so maybe it can be used to mitigate this problem, but the equality shouldn't be used unless you store a new job attribute `start_jiffies` (as soon as you convert to a human readable time, clock drift is an issue).

detailed log file from mom in during the start phase of the restart.
```
09/10/2015 11:30:34.580;02;   pbs_mom.12858;n/a;mom_server_update_stat;mom_server_update_stat: sending to server "state=free"
09/10/2015 11:30:34.580;02;   pbs_mom.12858;n/a;mom_server_update_stat;mom_server_update_stat: sending to server "jobs=28855.master(,session_id=11485)"
09/10/2015 11:30:34.580;02;   pbs_mom.12858;n/a;mom_server_update_stat;mom_server_update_stat: sending to server "varattr= "
09/10/2015 11:30:34.580;02;   pbs_mom.12858;n/a;mom_server_update_stat;mom_server_update_stat: sending to server "cpuclock=Fixed"
09/10/2015 11:30:34.580;02;   pbs_mom.12858;n/a;mom_server_update_stat;mom_server_update_stat: sending to server "macaddr=bc:30:5b:dc:ec:2e"
09/10/2015 11:30:34.580;08;   pbs_mom.12858;Job;read_tcp_reply;protocol: 4  version: 3  command:4  sock:14
09/10/2015 11:30:34.581;02;   pbs_mom.12858;n/a;mom_server_update_stat;status update successfully sent to master
09/10/2015 11:30:34.581;128;   pbs_mom.12854;Svr;mom_get_sample;proc_array load started
09/10/2015 11:30:34.582;02;   pbs_mom.12854;Svr;get_cpuset_pidlist;/dev/cpuset/torque/28855.master contains 5 PIDs
09/10/2015 11:30:34.582;02;   pbs_mom.12854;Svr;get_cpuset_pidlist;/dev/cpuset/torque contains 0 PIDs
09/10/2015 11:30:34.582;128;   pbs_mom.12854;n/a;mom_get_sample;proc_array loaded - nproc=5
09/10/2015 11:30:34.582;128;   pbs_mom.12854;n/a;cput_sum;pid2jobsid_map loop start - jobid = 28855.master
09/10/2015 11:30:34.582;128;   pbs_mom.12854;n/a;mem_sum;pid2jobsid_map loop start - jobid = 28855.master
09/10/2015 11:30:34.582;128;   pbs_mom.12854;n/a;resi_sum;proc_array loop start - jobid = 28855.master
09/10/2015 11:30:34.582;02;   pbs_mom.12854;Svr;get_cpuset_pidlist;/dev/cpuset/torque/28855.master contains 5 PIDs
09/10/2015 11:30:34.583;08;   pbs_mom.12854;Job;scan_non_child_tasks;found exited session 11485 for task 1 in job 28855.master
09/10/2015 11:30:34.583;08;   pbs_mom.12854;Svr;task_save;saving task in /var/spool/pbs/mom_priv/jobs/28855.master.TK
09/10/2015 11:30:34.583;128;   pbs_mom.12854;Svr;scan_for_exiting;searching for exiting jobs
09/10/2015 11:30:34.583;128;   pbs_mom.12854;Svr;update_job_based_on_tasks;Checking tasks for '28855.master'
09/10/2015 11:30:34.583;08;   pbs_mom.12854;Job;28855.master;job was terminated
09/10/2015 11:30:34.583;08;   pbs_mom.12854;Req;send_sisters;sending command KILL_JOB for job 28855.master (2)
09/10/2015 11:30:34.583;08;   pbs_mom.12854;Job;28855.master;no sisters contacted - setting job substate to EXITING
09/10/2015 11:30:34.583;08;   pbs_mom.12854;Job;28855.master;task is dead
09/10/2015 11:30:34.583;08;   pbs_mom.12854;Svr;task_save;saving task in /var/spool/pbs/mom_priv/jobs/28855.master.TK
09/10/2015 11:30:34.583;08;   pbs_mom.12854;Job;kill_job;mother_superior_cleanup: sending signal 9, "KILL" to job 28855.master, reason: local task termination detected
09/10/2015 11:30:34.583;08;   pbs_mom.12854;Job;28855.master;kill_job done (killed 0 processes)
09/10/2015 11:30:34.583;128;   pbs_mom.12854;Svr;mother_superior_cleanup;calling mom_open_socket_to_jobs_server
09/10/2015 11:30:34.583;128;   pbs_mom.12854;Svr;preobit_preparation;top
09/10/2015 11:30:34.584;128;   pbs_mom.12854;Job;28855.master;epilog subtask created with pid 12860 - substate set to JOB_SUBSTATE
```

**Comment by itkovian** [comment]
This issue prohibits us from easily addressing #315.

**Comment by stdweird** [comment]
actual evidence that btime is not fixed:
```
yum install torque-debuginfo -y
echo "print linux_time" > /root/pbsmomgdb
gdb -p `cat /var/spool/pbs/mom_priv/mom.lock` --batch --command=/root/pbsmomgdb |grep =| sed "s/.*=/mom_b/"
grep btime /proc/stat
```
and you get output like
```
mom_b 1438356139
btime 1438355912
```

**Comment by dbeer** [comment]
I'm closing this as it was addressed by pull request #325
